Fixes an issue with the ActiveRecord::ConnectionAdapters::SchemaCache#deep_deduplicate method, which affects Ruby 2.6

elohanlon opened the pull request:
### Summary

Ruby 2.6 and 2.7 have slightly different implementations of the String#@- method. In Ruby 2.6, the receiver of the String#@- method is modified under certain circumstances. This was later identified as a bug (https://bugs.ruby-lang.org/issues/15926) and only fixed in Ruby 2.7.

Before the changes in this commit, the ActiveRecord::ConnectionAdapters::SchemaCache#deep_deduplicate method, which internally calls the String#@- method, could also modify an input string argument in Ruby 2.6 -- changing a tainted, unfrozen string into a tainted, frozen string.

This PR fixes: https://github.com/rails/rails/issues/43056

elohanlon committed:
Fixes ActiveRecord::ConnectionAdapters::SchemaCache#deep_deduplicate for Ruby 2.6

Ruby 2.6 and 2.7 have slightly different implementations of the String#@- method. In Ruby 2.6, the receiver of the String#@- method is modified under certain circumstances. This was later identified as a bug (https://bugs.ruby-lang.org/issues/15926) and only fixed in Ruby 2.7.

Before the changes in this commit, the ActiveRecord::ConnectionAdapters::SchemaCache#deep_deduplicate method, which internally calls the String#@- method, could also modify an input string argument in Ruby 2.6 -- changing a tainted, unfrozen string into a tainted, frozen string.

casperisfine commented:
Thanks for the patch, but Rails 6.1 only receive security patches: https://guides.rubyonrails.org/maintenance_policy.html

elohanlon commented:
@casperisfine Was that a very recent change to the maintenance policy document?  Less than a month ago, @rafaelfranca suggested that I open a Rails PR for bug https://github.com/rails/rails/issues/43056 (which this PR fixes).

Maybe it's a stretch for me to call this a security issue, but for what it's worth the bug results in unexpected characters in generated SQL queries (thereby creating malformed queries).  In my particular case, I have examples of it leading to the transformation of one of my table names -- `enabled_dynamic_fields` -- into `��amic_fields`, `�S�(�ields`, `�%�� &��`, and other variations.

casperisfine commented:
Hum, no it dates from the 7.0 release. But if @rafaelfranca think it's worth a backport, I'll review your PR.

casperisfine commented on activerecord/lib/active_record/connection_adapters/schema_cache.rb:
```diff
@@ -204,7 +204,16 @@ def deep_deduplicate(value)
             value.map { |i| deep_deduplicate(i) }
-          when String, Deduplicable
+          when String
+            if RUBY_VERSION.start_with?('2.6') && (value.tainted? || !value.frozen?)
```
Generally you don't want this type of version check on a hotpath. People on other Ruby version should pay the cost for this.

You can have something like:

```ruby
if RUBY_VERSION.start_with?('2.6')
  def deep_deduplicate(value)
    # version handling the bug
  end
else
  def deep_deduplicate(value)
    # normal version
  end
end

casperisfine replied:
Also I'm not sure about the `(value.tainted? || !value.frozen?)`. Based on the test your wrote, the bug happens when `value.tained? && !value.frozen?`. 

Tainted strings are rare enough that just checking `tainted?` should be enough I think.

elohanlon commented:
Thank you for the review!  I just pushed a commit that addresses your feedback.

You're right that I've only seen the bug for `value.tained? && !value.frozen?`.  The reason I originally included the standalone `!value.frozen?` case was because the referenced ruby bug (https://bugs.ruby-lang.org/issues/15926) described a case where an untainted, frozen string was also modified.  I'm not seeing evidence of that in the test though, so glad to remove.

Does the updated PR look okay?

casperisfine commented:
Yep, the diff looks good now. It's just missing a CHANGELOG, and commits must be squashed. After that it's 👌

elohanlon commented:
CHANGELOG updated and commits squashed!

byroot commented:
Arf, rubocop doesn't like single quote. Sorry, after that I can merge.

elohanlon commented:
Fixed!

elohanlon commented:
Thank you for merging this!

elohanlon commented:
Will do as soon as possible. 👍

rafaelfranca commented:
This PR break tests for Ruby 2.5. What is the expected behavior in the version of Ruby?

elohanlon commented:
@rafaelfranca Thank you for catching that.  I originally found a bug while running Ruby 2.6 (but not 2.7) and scoped the fix to 2.6 in this PR, and added a new test, but that new test failure is indicating that the bug affects Ruby 2.5 as well.  To answer your question: The expected behavior is that the new test should pass in all versions of Ruby that are compatible with Rails 6, including 2.5.

In my code, I have a line like this:

```
if RUBY_VERSION.start_with?("2.6")
```

But I can expand the fix to ruby 2.5 by just changing that to:

```
if RUBY_VERSION.start_with?("2.5") || RUBY_VERSION.start_with?("2.6")
```

How would you like me to move forward?  Should I push another commit to this PR, or would it be better for me to open a new PR?

rafaelfranca commented:
Ah. I just applied a patch to make sure Ruby 2.5 has the same behavior of 2.6. Thank you for confirming.

elohanlon commented:
Thank you for taking care of that!
